connect-inject/cleanup_resource.go

```diff
@@ -0,0 +1,231 @@
+package connectinject
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	capi "github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-hclog"
+	"golang.org/x/net/context"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+// CleanupResource implements Resource and is used to clean up Consul service
+// instances that weren't deregistered when their pods were deleted.
+// Usually the preStop hook in the pods handles this but during a force delete
+// or OOM the preStop hook doesn't run.
+type CleanupResource struct {
+	Log              hclog.Logger
+	KubernetesClient kubernetes.Interface
+	// ConsulClient points at the agent on the same node as this pod.
+	ConsulClient    *capi.Client
+	ReconcilePeriod time.Duration
+	Ctx             context.Context
+	// ConsulScheme is the scheme to use when making API calls to Consul,
+	// i.e. "http" or "https".
+	ConsulScheme string
+	// ConsulPort is the port to make HTTP API calls to Consul agents on.
+	ConsulPort string
+
+	lock sync.Mutex
+}
+
+// Run starts the long-running Reconcile loop that runs on a timer.
```

Comment:
This is a slight change to the `Run` function from the health check resource (https://github.com/hashicorp/consul-k8s/blob/master/connect-inject/health_check_resource.go#L54-L76).

1. I refactored the loop so we don't have `h.Reconcile()` called twice. I believe they're logically equivalent but please double check.
2. `reconcile()` here doesn't return an error. I did this because we're already logging all the places there are errors inside `reconcile()` with a lot of good context as to why the error is occurring and so logging it again here just results in duplicate error logs.

I plan to contribute these changes back to `HealthCheckResource` in a separate PR.

Reply:
Great idea on point 3!